Mathlib/GameTheory/Auction/Secondprice.lean

```diff
@@ -0,0 +1,203 @@
+import Mathlib.Data.Real.Basic
+
+/-!
+# Auction Theory
+
+This file formalizes core concepts and results in auction theory. It includes the definitions of
+first-price and second-price auctions, as well as several fundamental results and helping lemmas.
+
+## Main Definitions
+
+- `Auction`: Defines an auction with bidders and their valuations.
+- `maxBid`: Function that computes the highest bid given a bidding function.
+- `winner`: Identifies the winner of the auction as the bidder with the highest bid.
+- `maxBidExcluding`: Function that computes the highest bid excluding a given participant.
+- `dominant`: Defines a dominant strategy in the context of a auction.
+- `Secondprice`: Computes the second highest bid in the auction.
+- `Secondprice.utility`: Computes the utility of each bidder based on the outcome of the auction.
+- `Firstprice.utility`: Computes the utility for a first price auction.
+
+## Main Results
+
+- `utility_nneg`: Utility is non-negative if the bid equals the valuation in second price auction.
+- `valuation_is_dominant`: Bidding one's valuation is a dominant strategy in second price auction.
+- `first_price_has_no_dominant_strategy`: There is no dominant strategy in first price auction.
+
+## Helping Lemmas
+
+- `eq_winner_of_bid_gt`: If `i`'s bid is higher than all other bids, then `i` wins.
+- `exists_maxBid`: There exists a participant whose bid matches the highest bid
+- `bid_winner_eq_maxBid`: The winner's bid is the highest.
+- `winnerbid_ge_second`: The winner's bid is at least the second highest bid.
+- `Secondprice.utility_winner`: If `i` wins, utility is the valuation minus the second highest bid.
+- `Secondprice.utility_loser`: If `i` does not win, their utility is 0.
+- `bid_le_bid_winner`: The winner's bid is greater than or equal to all other bids.
+- `maxBidExcluding_eq_maxBid_if_loser`: If `i` does not win, the highest bid excluding `i` equals
+   the highest bid.
+- `Firstprice.utility_winner`: If `i` wins in a first price auction utility is valuation minus bid.
+- `Firstprice.utility_loser`: If `i` does not win in a first price auction, utility is 0.
+
+## Implementation Notes
+
+The structure and functions assume the existence of multiple bidders to allow for meaningful
+auction dynamics. Definitions like `winner` and `maxBid` make use of Lean's `Finset` and `Classical`
+logic to handle potential non-constructive cases effectively.
+
+## References
+
+* [T. Roughgarden, *Twenty lectures on Algorithmic Game Theory*][roughgarden2016]
+
+## Tags
+
+auction, game theory, economics, bidding, valuation
+-/
+
+open Classical
+
+/-!### Structure Definition -/
+
+/-- The Auction structure is set with components including: -/
+structure Auction where
+   /-- A set of participants.-/
+   I : Type*
+   /-- A `Fintype` instance. -/
+   hF : Fintype I
+   /-- An `Nontrivial` instance. -/
+   hI : Nontrivial I
+   /-- A function mapping each participant to their valuation. -/
+   v : I → ℝ
```

Comment:
valuation functions appear in many places in game theory, outside auctions, including voting theory, and fair division theory. It might be useful from a mathlib perspective to factor out valuation functions and the various properties they could have. A more general valuation function would map each participant and some item from a type of items to a real number. Even in auction theory, there are auctions on bundles of items. A general form of auction functions would look like `AgentType → Set (ItemType) → ℝ`. Again take this with a pinch of salt. In mathlib a general definition that can be more widely used is helpful. Further this also belongs in the aforementioned `Basic.lean` file

Reply:
Yes,  shreyas! This is just what I am planned to do. I have already finished the formalization of Myerson's Lemma which proves to be a more generalize lemma and also widely used in many mechanism design not just auction. I am planned to do so after the successful merge of the formalization of First-price and Second-price auction.